File: perftest/simulations/src/test/scala/org/projectnessie/perftest/gatling/BaseParams.scala
/*
 * Copyright (C) 2020 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.projectnessie.perftest.gatling

/** Common parameters for Nessie Gatling simulations. */
trait BaseParams {

  /** The number of users to simulate. System property: `sim.users`, defaults to `1`. */
  def numUsers: Int

  /** The operation-rate in ops-per-second, `0` means "as fast as possible".
    * Implemented with Gatling's [[io.gatling.core.structure.Pauses#pace(Duration)]]
    * function, which acts like a "sleep()" after each commit.
    * System property: `sim.rate`, defaults to `0`.
    */
  def opRate: Double

  /** Arbitrary note added as a Prometheus tag, used to distinguish metrics from different runs in e.g. Grafana. System property: `sim.note`, defaults to `""`. */
  def note: String

  def asPrintableString(): String = {
    s"""
    |Simulation parameters:
    |   note:           $note
    |   num-users:      $numUsers
    |   op-rate:        $opRate
    |""".stripMargin
  }
}

case class BaseParamsImpl(
    override val numUsers: Int,
    override val opRate: Double,
    override val note: String
) extends BaseParams

object BaseParams {
  def fromSystemProperties(): BaseParams = {
    val numUsers: Int = Integer.getInteger("sim.users", 1).toInt
    val opRate: Double = System.getProperty("sim.rate", "0").toDouble
    val note: String = System.getProperty("sim.note", "")

    BaseParamsImpl(
      numUsers,
      opRate,
      note
    )
  }
}

File: servers/services-bench/src/jmh/java/org/projectnessie/services/BaseParams.java
/*
 * Copyright (C) 2023 Dremio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.projectnessie.services;

import static org.projectnessie.versioned.storage.common.logic.Logics.repositoryLogic;

import java.util.HashSet;
import java.util.ServiceLoader;
import java.util.Set;
import org.projectnessie.versioned.VersionStore;
import org.projectnessie.versioned.storage.common.config.StoreConfig;
import org.projectnessie.versioned.storage.common.persist.Backend;
import org.projectnessie.versioned.storage.common.persist.Persist;
import org.projectnessie.versioned.storage.common.persist.PersistFactory;
import org.projectnessie.versioned.storage.testextension.BackendTestFactory;
import org.projectnessie.versioned.storage.versionstore.VersionStoreImpl;

abstract class BaseParams {
  public static final String DEFAULT_BRANCH_NAME = "main";

  Backend backend;
  VersionStore versionStore;
  BackendTestFactory backendTestFactory;

  protected void init(String backendName) throws Exception {
    Set<String> known = new HashSet<>();
    for (BackendTestFactory candidate : ServiceLoader.load(BackendTestFactory.class)) {
      String name = candidate.getName();
      known.add(name);
      if (backendName.equals(name)) {
        backendTestFactory = candidate;
        break;
      }
    }
    if (backendTestFactory == null) {
      throw new IllegalArgumentException(
          "Could not find backend named " + backendName + ", known backends: " + known);
    }

    backendTestFactory.start();

    backend = backendTestFactory.createNewBackend();
    backend.setupSchema();
    PersistFactory factory = backend.createFactory();
    Persist persist = factory.newPersist(StoreConfig.Adjustable.empty());
    repositoryLogic(persist).initialize(DEFAULT_BRANCH_NAME);
    versionStore = new VersionStoreImpl(persist);
  }

  protected void tearDown() throws Exception {
    if (backend != null) {
      try {
        backend.close();
      } finally {
        backend = null;
      }
    }
    if (backendTestFactory != null) {
      try {
        backendTestFactory.stop();
      } finally {
        backendTestFactory = null;
      }
    }
  }
}
